File: grpc_api/server/src/auth/token.rs
use tonic::{Request, Response, Status};
use chrono::{Utc, TimeZone};
use uuid::Uuid;
use bbthings_database::Auth;
use bbthings_grpc_proto::auth::token::token_service_server::TokenService;
use bbthings_grpc_proto::auth::token::{
    TokenSchema, AuthToken, AccessId, UserId, AuthTokenCreate, TokenUpdate,
    TokenTime, TokenRangeSingle, TokenRangeDouble,
    TokenReadResponse, TokenListResponse, TokenCreateResponse, AuthTokenCreateResponse, 
    TokenUpdateResponse, TokenChangeResponse
};
use crate::common::validator::{AuthValidator, ValidatorKind};
use crate::common::utility::handle_error;

pub struct TokenServer {
    auth_db: Auth,
    validator_flag: bool
}

impl TokenServer {
    pub fn new(auth_db: Auth) -> Self {
        TokenServer {
            auth_db,
            validator_flag: false
        }
    }
    pub fn new_with_validator(auth_db: Auth) -> Self {
        TokenServer {
            auth_db,
            validator_flag: true
        }
    }
}

#[tonic::async_trait]
impl TokenService for TokenServer {

    async fn read_access_token(&self, request: Request<AccessId>)
        -> Result<Response<TokenReadResponse>, Status>
    {
        self.validate(request.extensions(), ValidatorKind::Root).await?;
        let request = request.into_inner();
        let result = self.auth_db.read_access_token(request.access_id).await;
        let result = match result {
            Ok(value) => Some(value.into()),
            Err(e) => return Err(handle_error(e))
        };
        Ok(Response::new(TokenReadResponse { result }))
    }

    async fn list_auth_token(&self, request: Request<AuthToken>)
        -> Result<Response<TokenListResponse>, Status>
    {
        self.validate(request.extensions(), ValidatorKind::Root).await?;
        let request = request.into_inner();
        let result = self.auth_db.list_auth_token(&request.auth_token).await;
        let results = match result {
            Ok(value) => value.into_iter().map(|e| e.into()).collect(),
            Err(e) => return Err(handle_error(e))
        };
        Ok(Response::new(TokenListResponse { results }))
    }

    async fn list_token_by_user(&self, request: Request<UserId>)
        -> Result<Response<TokenListResponse>, Status>
    {
        self.validate(request.extensions(), ValidatorKind::Root).await?;
        let request = request.into_inner();
        let result = self.auth_db.list_token_by_user(Uuid::from_slice(&request.user_id).unwrap_or_default()).await;
        let results = match result {
            Ok(value) => value.into_iter().map(|e| e.into()).collect(),
            Err(e) => return Err(handle_error(e))
        };
        Ok(Response::new(TokenListResponse { results }))
    }

    async fn list_token_by_created_earlier(&self, request: Request<TokenTime>)
        -> Result<Response<TokenListResponse>, Status>
    {
        self.validate(request.extensions(), ValidatorKind::Root).await?;
        let request = request.into_inner();
        let result = self.auth_db.list_token_by_created_earlier(
            Utc.timestamp_nanos(request.timestamp * 1000),
            request.user_id.map(|id| Uuid::from_slice(&id).unwrap_or_default())
        ).await;
        let results = match result {
            Ok(value) => value.into_iter().map(|e| e.into()).collect(),
            Err(e) => return Err(handle_error(e))
        };
        Ok(Response::new(TokenListResponse { results }))
    }

    async fn list_token_by_created_later(&self, request: Request<TokenTime>)
        -> Result<Response<TokenListResponse>, Status>
    {
        self.validate(request.extensions(), ValidatorKind::Root).await?;
        let request = request.into_inner();
        let result = self.auth_db.list_token_by_created_later(
            Utc.timestamp_nanos(request.timestamp * 1000),
            request.user_id.map(|id| Uuid::from_slice(&id).unwrap_or_default())
        ).await;
        let results = match result {
            Ok(value) => value.into_iter().map(|e| e.into()).collect(),
            Err(e) => return Err(handle_error(e))
        };
        Ok(Response::new(TokenListResponse { results }))
    }

    async fn list_token_by_created_range(&self, request: Request<TokenRangeSingle>)
        -> Result<Response<TokenListResponse>, Status>
    {
        self.validate(request.extensions(), ValidatorKind::Root).await?;
        let request = request.into_inner();
        let result = self.auth_db.list_token_by_created_range(
            Utc.timestamp_nanos(request.begin * 1000),
            Utc.timestamp_nanos(request.end * 1000),
            request.user_id.map(|id| Uuid::from_slice(&id).unwrap_or_default())
        ).await;
        let results = match result {
            Ok(value) => value.into_iter().map(|e| e.into()).collect(),
            Err(e) => return Err(handle_error(e))
        };
        Ok(Response::new(TokenListResponse { results }))
    }

    async fn list_token_by_expired_earlier(&self, request: Request<TokenTime>)
        -> Result<Response<TokenListResponse>, Status>
    {
        self.validate(request.extensions(), ValidatorKind::Root).await?;
        let request = request.into_inner();
        let result = self.auth_db.list_token_by_expired_earlier(
            Utc.timestamp_nanos(request.timestamp * 1000),
            request.user_id.map(|id| Uuid::from_slice(&id).unwrap_or_default())
        ).await;
        let results = match result {
            Ok(value) => value.into_iter().map(|e| e.into()).collect(),
            Err(e) => return Err(handle_error(e))
        };
        Ok(Response::new(TokenListResponse { results }))
    }

    async fn list_token_by_expired_later(&self, request: Request<TokenTime>)
        -> Result<Response<TokenListResponse>, Status>
    {
        self.validate(request.extensions(), ValidatorKind::Root).await?;
        let request = request.into_inner();
        let result = self.auth_db.list_token_by_expired_later(
            Utc.timestamp_nanos(request.timestamp * 1000),
            request.user_id.map(|id| Uuid::from_slice(&id).unwrap_or_default())
        ).await;
        let results = match result {
            Ok(value) => value.into_iter().map(|e| e.into()).collect(),
            Err(e) => return Err(handle_error(e))
        };
        Ok(Response::new(TokenListResponse { results }))
    }

    async fn list_token_by_expired_range(&self, request: Request<TokenRangeSingle>)
        -> Result<Response<TokenListResponse>, Status>
    {
        self.validate(request.extensions(), ValidatorKind::Root).await?;
        let request = request.into_inner();
        let result = self.auth_db.list_token_by_expired_range(
            Utc.timestamp_nanos(request.begin * 1000),
            Utc.timestamp_nanos(request.end * 1000),
            request.user_id.map(|id| Uuid::from_slice(&id).unwrap_or_default())
        ).await;
        let results = match result {
            Ok(value) => value.into_iter().map(|e| e.into()).collect(),
            Err(e) => return Err(handle_error(e))
        };
        Ok(Response::new(TokenListResponse { results }))
    }

    async fn list_token_by_range(&self, request: Request<TokenRangeDouble>)
        -> Result<Response<TokenListResponse>, Status>
    {
        self.validate(request.extensions(), ValidatorKind::Root).await?;
        let request = request.into_inner();
        let result = self.auth_db.list_token_by_range(
            Utc.timestamp_nanos(request.begin_1 * 1000),
            Utc.timestamp_nanos(request.end_1 * 1000),
            Utc.timestamp_nanos(request.begin_2 * 1000),
            Utc.timestamp_nanos(request.end_2 * 1000),
            request.user_id.map(|id| Uuid::from_slice(&id).unwrap_or_default())
        ).await;
        let results = match result {
            Ok(value) => value.into_iter().map(|e| e.into()).collect(),
            Err(e) => return Err(handle_error(e))
        };
        Ok(Response::new(TokenListResponse { results }))
    }

    async fn create_access_token(&self, request: Request<TokenSchema>)
        -> Result<Response<TokenCreateResponse>, Status>
    {
        self.validate(request.extensions(), ValidatorKind::Root).await?;
        let request = request.into_inner();
        let result = self.auth_db.create_access_token(
            Uuid::from_slice(&request.user_id).unwrap_or_default(),
            &request.auth_token,
            Utc.timestamp_nanos(request.expired * 1000),
            request.ip.as_slice()
        ).await;
        let (access_id, refresh_token, auth_token) = match result {
            Ok(value) => value,
            Err(e) => return Err(handle_error(e))
        };
        Ok(Response::new(TokenCreateResponse { access_id, refresh_token, auth_token }))
    }

    async fn create_auth_token(&self, request: Request<AuthTokenCreate>)
        -> Result<Response<AuthTokenCreateResponse>, Status>
    {
        self.validate(request.extensions(), ValidatorKind::Root).await?;
        let request = request.into_inner();
        let result = self.auth_db.create_auth_token(
            Uuid::from_slice(&request.user_id).unwrap_or_default(),
            Utc.timestamp_nanos(request.expire * 1000),
            request.ip.as_slice(),
            request.number as usize
        ).await;
        let tokens = match result {
            Ok(value) => value.into_iter()
                .map(|t| TokenCreateResponse {
                    access_id: t.0,
                    refresh_token: t.1,
                    auth_token: t.2
                }).collect(),
            Err(e) => return Err(handle_error(e))
        };
        Ok(Response::new(AuthTokenCreateResponse { tokens }))
    }

    async fn update_access_token(&self, request: Request<TokenUpdate>)
        -> Result<Response<TokenUpdateResponse>, Status>
    {
        self.validate(request.extensions(), ValidatorKind::Root).await?;
        let request = request.into_inner();
        let result = self.auth_db.update_access_token(
            request.access_id.unwrap_or_default(),
            request.expire.map(|s| Utc.timestamp_nanos(s * 1000)),
            request.ip.as_deref()
        ).await;
        let refresh_token = match result {
            Ok(value) => value,
            Err(e) => return Err(handle_error(e))
        };
        Ok(Response::new(TokenUpdateResponse { refresh_token }))
    }

    async fn update_auth_token( &self, request: Request<TokenUpdate>)
        -> Result<Response<TokenUpdateResponse>, Status>
    {
        self.validate(request.extensions(), ValidatorKind::Root).await?;
        let request = request.into_inner();
        let result = self.auth_db.update_auth_token(
            request.auth_token.unwrap_or_default().as_ref(),
            request.expire.map(|s| Utc.timestamp_nanos(s * 1000)),
            request.ip.as_deref()
        ).await;
        let refresh_token = match result {
            Ok(value) => value,
            Err(e) => return Err(handle_error(e))
        };
        Ok(Response::new(TokenUpdateResponse { refresh_token }))
    }

    async fn delete_access_token(&self, request: Request<AccessId>)
        -> Result<Response<TokenChangeResponse>, Status>
    {
        self.validate(request.extensions(), ValidatorKind::Root).await?;
        let request = request.into_inner();
        let result = self.auth_db.delete_access_token(request.access_id).await;
        match result {
            Ok(_) => (),
            Err(e) => return Err(handle_error(e))
        };
        Ok(Response::new(TokenChangeResponse { }))
    }

    async fn delete_auth_token(&self, request: Request<AuthToken>)
        -> Result<Response<TokenChangeResponse>, Status>
    {
        self.validate(request.extensions(), ValidatorKind::Root).await?;
        let request = request.into_inner();
        let result = self.auth_db.delete_auth_token(&request.auth_token).await;
        match result {
            Ok(_) => (),
            Err(e) => return Err(handle_error(e))
        };
        Ok(Response::new(TokenChangeResponse { }))
    }

    async fn delete_token_by_user(&self, request: Request<UserId>)
        -> Result<Response<TokenChangeResponse>, Status>
    {
        self.validate(request.extensions(), ValidatorKind::Root).await?;
        let request = request.into_inner();
        let result = self.auth_db.delete_token_by_user(Uuid::from_slice(&request.user_id).unwrap_or_default()).await;
        match result {
            Ok(_) => (),
            Err(e) => return Err(handle_error(e))
        };
        Ok(Response::new(TokenChangeResponse { }))
    }

}

impl AuthValidator for TokenServer {

    fn validator_flag(&self) -> bool {
        self.validator_flag
    }

    fn auth_db(&self) ->  &Auth {
        &self.auth_db
    }

}

File: grpc_api/client_web/src/auth/token.js
import pb_token from "../proto/auth/token_grpc_web_pb.js";
import {
    metadata,
    base64_to_uuid_hex,
    uuid_hex_to_base64,
    base64_to_bytes,
    bytes_to_base64
} from "../common/utility.js";


/**
 * @typedef {(string|Uint8Array)} Uuid
 */

/**
 * @typedef {Object} ServerConfig
 * @property {string} address
 * @property {?string} auth_token
 */

/**
 * @typedef {Object} AccessId
 * @property {number} access_id
 */

/**
 * @typedef {Object} AuthToken
 * @property {string} auth_token
 */

/**
 * @typedef {Object} UserId
 * @property {Uuid} id
 */

/**
 * @typedef {Object} TokenSchema
 * @property {number} access_id
 * @property {Uuid} user_id
 * @property {string} refresh_token
 * @property {string} auth_token
 * @property {Date} created
 * @property {Date} expired
 * @property {number[]|Uint8Array} ip
 */

/**
 * @param {*} r 
 * @returns {TokenSchema}
 */
function get_token_schema(r) {
    return {
        access_id: r.accessId,
        user_id: base64_to_uuid_hex(r.userId),
        refresh_token: r.refreshToken,
        auth_token: r.authToken,
        expired: new Date(r.expired / 1000),
        ip: base64_to_bytes(r.ip)
    };
}

/**
 * @param {*} r 
 * @returns {TokenSchema[]}
 */
function get_token_schema_vec(r) {
    return r.map((v) => {return get_token_schema(v)});
}

/**
 * @typedef {Object} TokenEarlier
 * @property {Date} earlier
 * @property {?Uuid} user_id
 */

/**
 * @typedef {Object} TokenLater
 * @property {Date} later
 * @property {?Uuid} user_id
 */

/**
 * @typedef {Object} TokenRange
 * @property {Date} begin
 * @property {Date} end
 * @property {?Uuid} user_id
 */

/**
 * @typedef {Object} TokenRanges
 * @property {Date} created_begin
 * @property {Date} created_end
 * @property {Date} expired_begin
 * @property {Date} expired_end
 * @property {?Uuid} user_id
 */

/**
 * @typedef {Object} AuthTokenCreate
 * @property {Uuid} user_id
 * @property {Date} expired
 * @property {number[]|Uint8Array} ip
 * @property {number} number
 */

/**
 * @typedef {Object} TokenUpdate
 * @property {?number} access_id
 * @property {?string} refresh_token
 * @property {?string} auth_token
 * @property {?Date} expired
 * @property {?number[]|Uint8Array} ip
 */

/**
 * @typedef {Object} TokenCreateResponse
 * @property {number} access_id
 * @property {string} refresh_token
 * @property {string} auth_token
 */

/**
 * @param {*} r 
 * @returns {TokenCreateResponse}
 */
function get_token_create_response(r) {
    return {
        access_id: r.accessId,
        refresh_token: r.refreshToken,
        auth_token: r.authToken,
    };
}

/**
 * @param {*} r 
 * @returns {TokenCreateResponse[]}
 */
function get_token_create_response_vec(r) {
    return r.map((v) => {return get_token_create_response(v)});
}

/**
 * @typedef {Object} TokenUpdateResponse
 * @property {string} refresh_token
 * @property {string} auth_token
 */

/**
 * @param {*} r 
 * @returns {TokenUpdateResponse}
 */
function get_token_update_response(r) {
    return {
        refresh_token: r.refreshToken,
        auth_token: r.authToken,
    };
}


/**
 * Read an access token by access id
 * @param {ServerConfig} config Auth server config: address, auth_token
 * @param {AccessId} request access id: access_id
 * @returns {Promise<TokenSchema>} token schema: access_id, user_id, refresh_token, auth_token, created, expired, ip
 */
export async function read_access_token(config, request) {
    const client = new pb_token.TokenServicePromiseClient(config.address, null, null);
    const accessId = new pb_token.AccessId();
    accessId.setAccessId(request.access_id);
    return client.readAccessToken(accessId, metadata(config.auth_token))
        .then(response => get_token_schema(response.toObject().result));
}

/**
 * Read tokens by auth token
 * @param {ServerConfig} config Auth server config: address, auth_token
 * @param {AuthToken} request auth token: auth_token
 * @returns {Promise<TokenSchema[]>} token schema: access_id, user_id, refresh_token, auth_token, created, expired, ip
 */
export async function list_auth_token(config, request) {
    const client = new pb_token.TokenServicePromiseClient(config.address, null, null);
    const authToken = new pb_token.AuthToken();
    authToken.setAuthToken(request.auth_token);
    return client.listAuthToken(authToken, metadata(config.auth_token))
        .then(response => get_token_schema_vec(response.toObject().resultsList));
}

/**
 * Read tokens by user id
 * @param {ServerConfig} config Auth server config: address, auth_token
 * @param {UserId} request user id: id
 * @returns {Promise<TokenSchema[]>} token schema: access_id, user_id, refresh_token, auth_token, created, expired, ip
 */
export async function list_token_by_user(config, request) {
    const client = new pb_token.TokenServicePromiseClient(config.address, null, null);
    const userId = new pb_token.UserId();
    userId.setUserId(uuid_hex_to_base64(request.id));
    return client.listTokenByUser(userId, metadata(config.auth_token))
        .then(response => get_token_schema_vec(response.toObject().resultsList));
}

/**
 * Read tokens by earlier created time
 * @param {ServerConfig} config Auth server config: address, auth_token
 * @param {TokenEarlier} request token earlier: earlier, user_id
 * @returns {Promise<TokenSchema[]>} token schema: access_id, user_id, refresh_token, auth_token, created, expired, ip
 */
export async function list_token_by_created_earlier(config, request) {
    const client = new pb_token.TokenServicePromiseClient(config.address, null, null);
    const tokenTime = new pb_token.TokenTime();
    tokenTime.setTimestamp(request.earlier.valueOf() * 1000);
    if (request.user_id) {
        tokenTime.setUserId(uuid_hex_to_base64(request.user_id));
    }
    return client.listTokenByCreatedEarlier(tokenTime, metadata(config.auth_token))
        .then(response => get_token_schema_vec(response.toObject().resultsList));
}

/**
 * Read tokens by later created time
 * @param {ServerConfig} config Auth server config: address, auth_token
 * @param {TokenLater} request token later: later, user_id
 * @returns {Promise<TokenSchema[]>} token schema: access_id, user_id, refresh_token, auth_token, created, expired, ip
 */
export async function list_token_by_created_later(config, request) {
    const client = new pb_token.TokenServicePromiseClient(config.address, null, null);
    const tokenTime = new pb_token.TokenTime();
    tokenTime.setTimestamp(request.later.valueOf() * 1000);
    if (request.user_id) {
        tokenTime.setUserId(uuid_hex_to_base64(request.user_id));
    }
    return client.listTokenByCreatedLater(tokenTime, metadata(config.auth_token))
        .then(response => get_token_schema_vec(response.toObject().resultsList));
}

/**
 * Read tokens by created time range
 * @param {ServerConfig} config Auth server config: address, auth_token
 * @param {TokenRange} request token range: begin, end, user_id
 * @returns {Promise<TokenSchema[]>} token schema: access_id, user_id, refresh_token, auth_token, created, expired, ip
 */
export async function list_token_by_created_range(config, request) {
    const client = new pb_token.TokenServicePromiseClient(config.address, null, null);
    const tokenRange = new pb_token.TokenRangeSingle();
    tokenRange.setBegin(request.begin.valueOf() * 1000);
    tokenRange.setEnd(request.end.valueOf() * 1000);
    if (request.user_id) {
        tokenRange.setUserId(uuid_hex_to_base64(request.user_id));
    }
    return client.listTokenByCreatedRange(tokenRange, metadata(config.auth_token))
        .then(response => get_token_schema_vec(response.toObject().resultsList));
}

/**
 * Read tokens by earlier expired time
 * @param {ServerConfig} config Auth server config: address, auth_token
 * @param {TokenEarlier} request token earlier: earlier, user_id
 * @returns {Promise<TokenSchema[]>} token schema: access_id, user_id, refresh_token, auth_token, created, expired, ip
 */
export async function list_token_by_expired_earlier(config, request) {
    const client = new pb_token.TokenServicePromiseClient(config.address, null, null);
    const tokenTime = new pb_token.TokenTime();
    tokenTime.setTimestamp(request.earlier.valueOf() * 1000);
    if (request.user_id) {
        tokenTime.setUserId(uuid_hex_to_base64(request.user_id));
    }
    return client.listTokenByExpiredEarlier(tokenTime, metadata(config.auth_token))
        .then(response => get_token_schema_vec(response.toObject().resultsList));
}

/**
 * Read tokens by later expired time
 * @param {ServerConfig} config Auth server config: address, auth_token
 * @param {TokenLater} request token later: later, user_id
 * @returns {Promise<TokenSchema[]>} token schema: access_id, user_id, refresh_token, auth_token, created, expired, ip
 */
export async function list_token_by_expired_later(config, request) {
    const client = new pb_token.TokenServicePromiseClient(config.address, null, null);
    const tokenTime = new pb_token.TokenTime();
    tokenTime.setTimestamp(request.later.valueOf() * 1000);
    if (request.user_id) {
        tokenTime.setUserId(uuid_hex_to_base64(request.user_id));
    }
    return client.listTokenByExpiredLater(tokenTime, metadata(config.auth_token))
        .then(response => get_token_schema_vec(response.toObject().resultsList));
}

/**
 * Read tokens by expired time range
 * @param {ServerConfig} config Auth server config: address, auth_token
 * @param {TokenRange} request token range: begin, end, user_id
 * @returns {Promise<TokenSchema[]>} token schema: access_id, user_id, refresh_token, auth_token, created, expired, ip
 */
export async function list_token_by_expired_range(config, request) {
    const client = new pb_token.TokenServicePromiseClient(config.address, null, null);
    const tokenRange = new pb_token.TokenRangeSingle();
    tokenRange.setBegin(request.begin.valueOf() * 1000);
    tokenRange.setEnd(request.end.valueOf() * 1000);
    if (request.user_id) {
        tokenRange.setUserId(uuid_hex_to_base64(request.user_id));
    }
    return client.listTokenByExpiredRange(tokenRange, metadata(config.auth_token))
        .then(response => get_token_schema_vec(response.toObject().resultsList));
}

/**
 * Read tokens by created and expired time range
 * @param {ServerConfig} config Auth server config: address, auth_token
 * @param {TokenRanges} request token ranges: begin, end, user_id
 * @returns {Promise<TokenSchema[]>} token schema: access_id, user_id, refresh_token, auth_token, created, expired, ip
 */
export async function list_token_by_range(config, request) {
    const client = new pb_token.TokenServicePromiseClient(config.address, null, null);
    const tokenRanges = new pb_token.TokenRangeDouble();
    tokenRanges.setBegin1(request.created_begin.valueOf() * 1000);
    tokenRanges.setEnd1(request.created_end.valueOf() * 1000);
    tokenRanges.setBegin2(request.expired_begin.valueOf() * 1000);
    tokenRanges.setEnd2(request.expired_end.valueOf() * 1000);
    if (request.user_id) {
        tokenRanges.setUserId(uuid_hex_to_base64(request.user_id));
    }
    return client.listTokenByRange(tokenRanges, metadata(config.auth_token))
        .then(response => get_token_schema_vec(response.toObject().resultsList));
}

/**
 * Create an access token
 * @param {ServerConfig} config Auth server config: address, auth_token
 * @param {TokenSchema} request token schema: user_id, auth_token, expired, expired, ip
 * @returns {Promise<TokenCreateResponse>} create response: access_id, refresh_token, auth_token
 */
export async function create_access_token(config, request) {
    const client = new pb_token.TokenServicePromiseClient(config.address, null, null);
    const tokenSchema = new pb_token.TokenSchema();
    tokenSchema.setUserId(uuid_hex_to_base64(request.user_id));
    tokenSchema.setAuthToken(request.auth_token);
    tokenSchema.setExpired(request.expired.valueOf() * 1000);
    tokenSchema.setIp(bytes_to_base64(request.ip));
    return client.createAccessToken(tokenSchema, metadata(config.auth_token))
        .then(response => get_token_create_response(response.toObject()));
}

/**
 * Create tokens with shared auth token
 * @param {ServerConfig} config Auth server config: address, auth_token
 * @param {AuthTokenCreate} request token schema: user_id, expired, ip, number
 * @returns {Promise<TokenCreateResponse[]>} create response: access_id, refresh_token, auth_token
 */
export async function create_auth_token(config, request) {
    const client = new pb_token.TokenServicePromiseClient(config.address, null, null);
    const authTokenCreate = new pb_token.AuthTokenCreate();
    authTokenCreate.setUserId(uuid_hex_to_base64(request.user_id));
    authTokenCreate.setExpired(request.expired.valueOf() * 1000);
    authTokenCreate.setIp(bytes_to_base64(request.ip));
    authTokenCreate.setNumber(request.number);
    return client.createAuthToken(authTokenCreate, metadata(config.auth_token))
        .then(response => get_token_create_response_vec(response.toObject().tokensList));
}

/**
 * Update an access token
 * @param {ServerConfig} config Auth server config: address, auth_token
 * @param {TokenUpdate} request token update: access_id, expired, ip
 * @returns {Promise<TokenUpdateResponse>} update response: refresh_token, auth_token
 */
export async function update_access_token(config, request) {
    const client = new pb_token.TokenServicePromiseClient(config.address, null, null);
    const tokenUpdate = new pb_token.TokenUpdate();
    tokenUpdate.setAccessId(request.access_id);
    if (request.expired instanceof Date) {
        tokenUpdate.setExpired(request.expired.valueOf() * 1000);
    }
    if (request.ip) {
        tokenUpdate.setIp(bytes_to_base64(request.ip));
    }
    return client.updateAccessToken(tokenUpdate, metadata(config.auth_token))
        .then(response => get_token_update_response(response.toObject()));
}

/**
 * Update all tokens with shared auth token
 * @param {ServerConfig} config Auth server config: address, auth_token
 * @param {TokenUpdate} request token update: auth_token, expired, ip
 * @returns {Promise<TokenUpdateResponse>} update response: refresh_token, auth_token
 */
export async function update_auth_token(config, request) {
    const client = new pb_token.TokenServicePromiseClient(config.address, null, null);
    const tokenUpdate = new pb_token.TokenUpdate();
    tokenUpdate.setAuthToken(request.auth_token);
    if (request.expired instanceof Date) {
        tokenUpdate.setExpired(request.expired.valueOf() * 1000);
    }
    if (request.ip) {
        tokenUpdate.setIp(bytes_to_base64(request.ip));
    }
    return client.updateAuthToken(tokenUpdate, metadata(config.auth_token))
        .then(response => get_token_update_response(response.toObject()));
}

/**
 * Delete an access token by access id
 * @param {ServerConfig} config Auth server config: address, auth_token
 * @param {AccessId} request access id: access_id
 * @returns {Promise<{}>} delete response
 */
export async function delete_access_token(config, request) {
    const client = new pb_token.TokenServicePromiseClient(config.address, null, null);
    const accessId = new pb_token.AccessId();
    accessId.setAccessId(request.access_id);
    return client.deleteAccessToken(accessId, metadata(config.auth_token))
        .then(response => response.toObject());
}

/**
 * Delete tokens by auth token
 * @param {ServerConfig} config Auth server config: address, auth_token
 * @param {AuthToken} request auth token: auth_token
 * @returns {Promise<{}>} delete response
 */
export async function delete_auth_token(config, request) {
    const client = new pb_token.TokenServicePromiseClient(config.address, null, null);
    const authToken = new pb_token.AuthToken();
    authToken.setAuthToken(request.auth_token);
    return client.deleteAuthToken(authToken, metadata(config.auth_token))
        .then(response => response.toObject());
}

/**
 * Delete tokens by user id
 * @param {ServerConfig} config Auth server config: address, auth_token
 * @param {UserId} request user id: id
 * @returns {Promise<{}>} delete response
 */
export async function delete_token_by_user(config, request) {
    const client = new pb_token.TokenServicePromiseClient(config.address, null, null);
    const userId = new pb_token.UserId();
    userId.setUserId(uuid_hex_to_base64(request.id));
    return client.deleteTokenByUser(userId, metadata(config.auth_token))
        .then(response => response.toObject());
}
